File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/FoodDaoTest.java
package ediger.diarynutrition.db;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.room.Room;
import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;

import ediger.diarynutrition.LiveDataTestUtil;
import ediger.diarynutrition.data.source.DiaryDatabase;
import ediger.diarynutrition.data.source.entities.Food;
import ediger.diarynutrition.data.source.dao.FoodDao;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;


@RunWith(AndroidJUnit4.class)
public class FoodDaoTest {

    static final Food FOOD = new Food("test food", 1, 1, 1, 1);

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    private DiaryDatabase mDatabase;

    private FoodDao mFoodDao;

    @Before
    public void initDb() {
        mDatabase = Room.inMemoryDatabaseBuilder(InstrumentationRegistry.getContext(),
                DiaryDatabase.class)
                .allowMainThreadQueries()
                .build();
        mFoodDao = mDatabase.foodDao();
    }

    @After
    public void closeDb() {
        mDatabase.close();
    }

    @Test
    public void insertFoodAndGetAll() throws Exception {
        mFoodDao.insertFood(FOOD);

        List<Food> foods = LiveDataTestUtil.getValue(mFoodDao.getAllFood());

        assertThat(foods.size(), is(1));

        Food food = foods.get(0);

        assertFood(food, foods.size(), FOOD.getName(), FOOD.getCal(), FOOD.getProt(),
                FOOD.getFat(), FOOD.getCarbo());
    }

    @Test
    public void deleteFoodById() throws Exception {
        mFoodDao.insertFood(FOOD);

        List<Food> foods = LiveDataTestUtil.getValue(mFoodDao.getAllFood());

        Food food = foods.get(0);

        int delCount = mFoodDao.deleteFoodById(food.getId());

        assertThat(delCount, is(1));

        foods = LiveDataTestUtil.getValue(mFoodDao.getAllFood());

        assertThat(foods.size(), is(0));

    }

    @Test
    public void insertAndUpdate() {
        mFoodDao.insertFood(FOOD);
        Food food = mFoodDao.getFood(1);
        food.setFavorite(1);
        food.setUser(1);
        mFoodDao.updateFood(food);

        food = mFoodDao.getFood(1);

        assertFood(food, 1, FOOD.getName(), FOOD.getCal(),
                FOOD.getProt(), FOOD.getFat(), FOOD.getCarbo());
        assertThat(food.getFavorite(), is(1));
        assertThat(food.getUser(), is(1));
    }

    @Test
    public void insertAndGetFavor() throws Exception {
        Food food = new Food(FOOD.getName(), FOOD.getCal(),
                FOOD.getProt(), FOOD.getFat(), FOOD.getCarbo());
        food.setFavorite(1);
        mFoodDao.insertFood(food);
        mFoodDao.insertFood(food);

        List<Food> foods = LiveDataTestUtil.getValue(mFoodDao.getFavorFood());
        assertThat(foods.size(), is(2));

        food = foods.get(0);

        assertFood(food, 1, FOOD.getName(), FOOD.getCal(),
                FOOD.getProt(), FOOD.getFat(), FOOD.getCarbo());
        assertThat(food.getFavorite(), is(1));
    }

    @Test
    public void insertAndGetUsr() throws Exception {
        Food food = new Food(FOOD.getName(), FOOD.getCal(),
                FOOD.getProt(), FOOD.getFat(), FOOD.getCarbo());
        food.setUser(1);
        mFoodDao.insertFood(food);
        mFoodDao.insertFood(food);

        List<Food> foods = LiveDataTestUtil.getValue(mFoodDao.getUserFood());
        assertThat(foods.size(), is(2));

        food = foods.get(0);

        assertFood(food, 1, FOOD.getName(), FOOD.getCal(),
                FOOD.getProt(), FOOD.getFat(), FOOD.getCarbo());
        assertThat(food.getFavorite(), is(0));
        assertThat(food.getUser(), is(1));
    }

    @Test
    public void insertAndSearchAll() throws Exception {
        mFoodDao.insertFood(FOOD);
        Food food = new Food("food", 1, 1, 1, 1);
        mFoodDao.insertFood(food);

        List<Food> foods = LiveDataTestUtil.getValue(mFoodDao.searchAllFood("%food%"));
        assertThat(foods.size(), is(2));

        foods = LiveDataTestUtil.getValue(mFoodDao.searchAllFood("%test%"));
        assertThat(foods.size(), is(1));

        assertThat(foods.get(0).getName(), is("test food"));
    }

    @Test
    public void insertAndSearchFavor() throws Exception {
        Food food = new Food("test food", 1, 1, 1, 1);
        food.setFavorite(1);
        mFoodDao.insertFood(food);
        food.setName("food");
        mFoodDao.insertFood(food);

        List<Food> foods = LiveDataTestUtil.getValue(mFoodDao.searchFavorFood("%food%"));
        assertThat(foods.size(), is(2));

        foods = LiveDataTestUtil.getValue(mFoodDao.searchFavorFood("%test%"));
        assertThat(foods.size(), is(1));

        assertThat(foods.get(0).getFavorite(), is(1));
        assertThat(foods.get(0).getName(), is("test food"));
    }

    @Test
    public void insertAndGetSummary() throws Exception {

    }

    static void assertFood(Food food, int id, String foodName,
                            float cal, float prot, float fat, float carbo) {
        assertThat(food, notNullValue());
        assertThat(food.getId(), is(id));
        assertThat(food.getName(), is(foodName));
        assertThat(food.getCal(), is(cal));
        assertThat(food.getProt(), is(prot));
        assertThat(food.getFat(), is(fat));
        assertThat(food.getCarbo(), is(carbo));
    }

}

File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/FoodDaoTest.kt
package ediger.diarynutrition.db


import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import ediger.diarynutrition.data.source.DiaryDatabase
import ediger.diarynutrition.data.source.entities.Food
import ediger.diarynutrition.getOrAwaitValue
import ediger.diarynutrition.observeForTesting
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsNull.notNullValue
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

fun Food.assertFood(original: Food) {
    assertThat(this, notNullValue())
    assertThat(this.name, `is`(original.name))
    assertThat(this.cal, `is`(original.cal))
    assertThat(this.prot, `is`(original.prot))
    assertThat(this.fat, `is`(original.fat))
    assertThat(this.carbo, `is`(original.carbo))
    assertThat(this.gi, `is`(original.gi))
    assertThat(this.favorite, `is`(original.favorite))
    assertThat(this.user, `is`(original.user))
    assertThat(this.verified, `is`(original.verified))
}

@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4::class)
class FoodDaoTest {

    @get:Rule
    var instantExecutorRule = InstantTaskExecutorRule()

    private lateinit var database: DiaryDatabase

    @Before
    fun initDb() {
        database = Room.inMemoryDatabaseBuilder(
                ApplicationProvider.getApplicationContext(),
                DiaryDatabase::class.java
        ).build()
    }

    @After
    fun closeDb() {
        database.close()
    }

    @Test
    fun insertFoodAndGetAll() = runBlockingTest {
        database.foodDao().insertFood(FOOD)
        database.foodDao().insertFood(FOOD_2)

        val foodList = database.foodDao().getAllFood().getOrAwaitValue()

        assertThat(foodList.size, `is`(2))

        foodList.find { it.name == FOOD.name }!!.assertFood(FOOD)
        foodList.find { it.name == FOOD_2.name }!!.assertFood(FOOD_2)
    }

    @Test
    fun deleteFoodById() = runBlockingTest {
        database.foodDao().insertFood(FOOD)

        var foodList = database.foodDao().getAllFood().getOrAwaitValue()
        val food = foodList.first()

        val delCount = database.foodDao().deleteFoodById(food.id)

        assertThat(delCount, `is`(1))

        foodList = database.foodDao().getAllFood().getOrAwaitValue()

        assertThat(foodList.size, `is`(0))
    }

    @Test
    fun updateFood() = runBlockingTest {
        database.foodDao().insertFood(FOOD)

        var foodList = database.foodDao().getAllFood().getOrAwaitValue()
        var food = foodList.first().apply {
            favorite = 1
            user = 1
        }

        database.foodDao().updateFood(food)

        foodList = database.foodDao().getAllFood().getOrAwaitValue()
        assertThat(foodList.size, `is`(1))
        food = foodList.first()

        assertThat(food, notNullValue())
        assertThat(food.favorite, `is`(1))
        assertThat(food.user, `is`(1))
    }

    @Test
    fun getFavorite() = runBlockingTest {
        database.foodDao().insertFood(FOOD_2)
        database.foodDao().insertFood(FOOD.apply { favorite = 1 })

        val foodList = database.foodDao().getFavorFood().getOrAwaitValue()
        assertThat(foodList.size, `is`(1))

        foodList.first().assertFood(FOOD.apply { favorite = 1 })
    }

    @Test
    fun getUserFood() = runBlockingTest {
        database.foodDao().insertFood(FOOD_2)
        database.foodDao().insertFood(FOOD.apply { user = 1 })

        val foodList = database.foodDao().getUserFood().getOrAwaitValue()
        assertThat(foodList.size, `is`(1))

        foodList.first().assertFood(FOOD.apply { user = 1 })
    }

    // Paging testing
    //fun insertAndSearchAll() {}
    //fun insertAndSearchFavorite() {}
    //fun insertAndSearchUserFood() {}

    companion object {
        @JvmField
        val FOOD = Food("test food", 1f, 1f, 1f, 1f)
        val FOOD_2 = Food("another one", 2f, 2f, 2f, 2f)
    }
}

